broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/RandomizedPartitionTransitionTest.java

```diff
@@ -0,0 +1,412 @@
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.framework;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import io.atomix.raft.RaftServer;
+import io.atomix.raft.RaftServer.Role;
+import io.camunda.zeebe.broker.system.partitions.PartitionTransitionContext;
+import io.camunda.zeebe.broker.system.partitions.PartitionTransitionStep;
+import io.camunda.zeebe.broker.system.partitions.impl.steps.StreamProcessorTransitionStep;
+import io.camunda.zeebe.engine.processing.streamprocessor.StreamProcessor;
+import io.camunda.zeebe.engine.processing.streamprocessor.StreamProcessorBuilder;
+import io.camunda.zeebe.util.health.HealthMonitor;
+import io.camunda.zeebe.util.sched.Actor;
+import io.camunda.zeebe.util.sched.ActorScheduler;
+import io.camunda.zeebe.util.sched.future.ActorFuture;
+import io.camunda.zeebe.util.sched.future.CompletableActorFuture;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import net.jqwik.api.Arbitraries;
+import net.jqwik.api.Arbitrary;
+import net.jqwik.api.Combinators;
+import net.jqwik.api.ForAll;
+import net.jqwik.api.GenerationMode;
+import net.jqwik.api.Property;
+import net.jqwik.api.Provide;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RandomizedPartitionTransitionTest {
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(RandomizedPartitionTransitionTest.class);
+
+  /**
+   * Verifies that during transitions at most one {@code StreamProcessor} is created. It sets up the
+   * following transition chain:
+   *
+   * <ol>
+   *   <li>Pausable dummy step
+   *   <li>{@code StreamProcessorTransitionStep}
+   * </ol>
+   *
+   * The first step is there to manipulate execution order. In particular, the step will wait for a
+   * countdown latch thus pausing transition execution. This in turn, allows scheduling successive
+   * transition which cancel their predecessors
+   *
+   * @param operations the operations to run
+   */
+  @Property(generation = GenerationMode.EXHAUSTIVE)
+  void atMostOneStreamProcessorIsRunningAtAnyTime(
+      @ForAll("testOperations") final List<TestOperation> operations) {
+    LOGGER.debug(
+        format(
+            "Testing property 'atMostOneStreamProcessorIsRunningAtAnyTime' on sequence %s",
+            operations));
+
+    final var actorScheduler = ActorScheduler.newActorScheduler().build();
+    actorScheduler.start();
+    try {
+      final var actor = new TestActor();
+      actorScheduler.submitActor(actor);
+
+      final var instanceTracker =
+          new PropertyAssertingInstanceTracker<StreamProcessor>() {
+            @Override
+            void assertProperties() {
+              assertThat(opened).describedAs("Active stream processors").hasSizeLessThan(2);
+            }
+          };
+
+      final var mockStreamProcessorBuilder = produceMockStreamProcessorBuilder(instanceTracker);
```

Comment:
> For me the value lies in the test being able to fail if either the transition logic is changed or if the step logic changes (e.g. someone modifying the conditions in prepare or transitionTo)

Shouldn't the unit test `StreamProcessorTransitionStepTest` take care of it? 

> What do you feel is gained by using the real StreamProcessor?

I don't think that we get anything by using real StreamProcessor also. For me using the real StreamProcessorStep with all these mocks make the test unnecessarily complicated with not much value. This test is about testing the behavior of `PartitionTransitionImpl`. Note that the bug was in the implementation of PartitionTransition.

Btw, I do not want to block this PR because of this. Ideally, I would like to see this test tests the behavior of `PartitionTransitionImpl`, and unit tests for each step tests take care of the testing each steps. If the current unit tests are not testing these scenarios, we should extend them.

Reply:
There is https://github.com/camunda-cloud/zeebe/pull/8101 to merge the fix and narrow test, and take the heat from this discussion.

I see the point you are making. I am still a little confused, though.

For me this test could be one of two things:
- as it is now, testing behavior of `PartitionTransitionImpl` and `StreamProcessorTransitionStep` together (and maybe test properties for other steps in the future)
- a test that only tests `PartitionTransitionsImpl` and uses no concrete step at all. But if it is a test like that, I feel it is more a contract test for the steps. And then the property, in my mind, should be that `prepareTransition` is called right before `transitionTo` and with the right arguments. And with that contract verified in this test, and with the unit tests for the step then you have full coverage.

I think both are valid approaches. But the second test would have a property that is relies on implementation details (i.e. which methods are called in which order) and earlier in this discussion you raised objections to this approach.

And testing only `PartitionTransitionsImpl` but verifying a property that fits better to a concrete implementation of a step, I don't see the value in this approach.

Happy to discuss this. Not sure what you have in mind and what I am missing.

Reply:
We can of course discuss this. But as I said, this is not a blocker for me. For me the blocker is the number of operations (as I mentioned in another comment, to figure out if 3 is enough or not).
Let's find some time to discuss this tomorrow and let's agree on something so that you can make progress.

Reply:
After the discussions @pihme convinced me to use the real steps here :slightly_smiling_face: One thing we discussed was to also add zbdb step in the tests as the conditions to install and remove it different from StreamProcessorStep. But you can also do it as a follow up.